rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+             – (-) Hard to discern visually
+         - `-`
+             - (+) Visually not intrusive
+         - `_`
+             – (-) Visually more intrusive
+         - `^`
+             – (-) Overlaps with derivation output syntax
+         - All of the following characters will interfere with some tooling:
+             - `!` - shells
+             - `"` - shells
+             - `#` - URLs
+             - `$` - shells
+             - `%` - URLs
+             - `&` - shells, URLs
+             - `+` - URLs
+             - `,` - natural language
+             - `/` - paths, URLs
+             - `:` - URLs
+             - `;` - shells
+             - `=` - URLs, Nix language
+             - `?` - URLs
+             - `@` - URLs, Nix language
+             - `\` - Windows paths
+     - (-) `default.nix` resolving needs specification:
+         - If for `import ./foo`, all of `./foo/default.nix{6,7,8}` exist, pick the one matching the version used by the evaluator, otherwise fail
+             - Then you'd have to specify a file using a different version explicitly
+
+   - Language versioning per "project" in a sidecar file
+       - (+) This would easily allow inheriting the language version across imports (obviating many specifications in this proposal)
+       - (-) There is currently no notion of "project" in the Nix language
+         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
+       - (-) Cannot be introduced gradually, particularly relevant for a large codebase like Nixpkgs
+       - (-) It would require a separate language to encode project metadata such as the language version
+           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
+           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
+             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata
+
+   </details>
+
+1. The following syntax is used for declaring the language version of a Nix expression:
+
+   ```
+   version \d*;
+   ```
+
+   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).
+
+   The syntax is open for bikeshedding.
+   Alternatives should be very short and self-describing.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Short and fairly self-descriptive
+   - (+) Not an invasive change
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - `use v\d*;`
+     - (+) Shorter
+     - (-) Does not explain much
+   - `Nix language version \d*`
+     - (+) Says it all
+     - (-) Very long
+   - `with import <language> \d*;`
+     - (+) Allows for forward compatibilty hacks such as better error messages
```

Comment:
Another benefit: less cost for «mostly compatible» implementations. E.g. syntax-highighting the version statement exactly in the beginning is annoying, here something mostly sensible is easy.

Reply:
Not sure how this is different from the forward compatibility hacks in general.

```suggestion
     - (+) Allows for forward compatibility hacks such as better error messages
```

Reply:
Forward compatibility hacks are something to do, here doing _nothing at all_ would be good enough for many things.

Reply:
This syntax would only be required for declaring a minimum evaluator version >6 to parse the expression. If you don't change the expression, there is nothing to do to begin with, as according to this proposal, every evaluator would default to version 6 anyway.

This specific argument is about what happens if your expression is written for v7 but your evaluator only supports v6.

Reply:
Complex version interaction rules mean that there will be eventually a lot of «v7» files that are actually compatible with both v6 and v7. Well, unless v7 ends up completely unrelated to v6 syntactically.

Reply:
I'm not sure if this is going in the right direction. *Any* change to the language that is not accounted for by version bumps may fail either catastrophically or subtly (which is even worse for reproducibility) when you feed newer expressions to older evaluators. This proposal alleviates exactly that risk of expressions using *seemingly* the same syntax being confused for meaning the same thing. Example: changed semantics of `fetchGit`

Reply:
Not all tools are evaluators.